src/Altinn.App.Core/Internal/Expressions/LayoutEvaluatorStateInitializer.cs

```diff
@@ -9,34 +11,94 @@ namespace Altinn.App.Core.Internal.Expressions;
     /// </summary>
-    public LayoutEvaluatorStateInitializer(IAppResources appResources, IOptions<FrontEndSettings> frontEndSettings)
+    public LayoutEvaluatorStateInitializer(
+        IAppResources appResources,
+        IOptions<FrontEndSettings> frontEndSettings,
+        ICachedFormDataAccessor dataAccessor
+    )
     {
         _appResources = appResources;
+        _dataAccessor = dataAccessor;
         _frontEndSettings = frontEndSettings.Value;
     }
 
     /// <summary>
     /// Initialize LayoutEvaluatorState with given Instance, data object and layoutSetId
     /// </summary>
-    public virtual Task<LayoutEvaluatorState> Init(
+    [Obsolete("Use the overload with ILayoutEvaluatorStateInitializer instead")]
+    public Task<LayoutEvaluatorState> Init(
         Instance instance,
         object data,
         string? layoutSetId,
         string? gatewayAction = null
     )
     {
         var layouts = _appResources.GetLayoutModel(layoutSetId);
+        var dataElement = instance.Data.Find(d => d.DataType == layouts.DefaultDataType.Id);
+        Debug.Assert(dataElement is not null);
         return Task.FromResult(
-            new LayoutEvaluatorState(new DataModel(data), layouts, _frontEndSettings, instance, gatewayAction)
+            new LayoutEvaluatorState(
+                new DataModel([KeyValuePair.Create(dataElement, data)]),
+                layouts,
+                _frontEndSettings,
+                instance,
+                gatewayAction
+            )
+        );
+    }
+
+    /// <inheritdoc />
+    public async Task<LayoutEvaluatorState> Init(
+        Instance instance,
+        string taskId,
+        string? gatewayAction = null,
+        string? language = null
+    )
+    {
+        var layouts = _appResources.GetLayoutModelForTask(taskId);
+
+        var defaultDataTypeId = layouts.DefaultDataType.Id;
+        var defaultDataElement = instance.Data.Find(d => d.DataType == defaultDataTypeId);
+        if (defaultDataElement is null)
+        {
+            throw new InvalidOperationException($"No data element found for data type {defaultDataTypeId}");
+        }
+
+        var dataTasks = new List<Task<KeyValuePair<DataElement, object>>>();
+        foreach (var dataType in layouts.GetExternalModelReferences())
+        {
+            dataTasks.AddRange(
+                instance
+                    .Data.Where(dataElement => dataElement.DataType == dataType)
+                    .Select(async dataElement =>
+                        KeyValuePair.Create(dataElement, await _dataAccessor.Get(instance, dataElement))
+                    )
+            );
+        }
+
+        var extraModels = await Task.WhenAll(dataTasks);
+
+        return new LayoutEvaluatorState(
+            new DataModel(
+                extraModels.Append(
+                    KeyValuePair.Create(defaultDataElement, await _dataAccessor.Get(instance, defaultDataElement))
+                )
+            ),
```

Comment:
The `DataModel` constructor crashes with `System.ArgumentException: An item with the same key has already been added.` if you reference the default data model explicitly in data model bindings or expressions somewhere, as it will then be present in `GetExternaModelReferences()` before being added again at the end.

This also causes the default model to be fetched twice so it should not be fixed by checking the already added models inside of the `DataModel` constructor. I suggest you add the default data type in the same way as the "external" ones to ensure uniqueness.

Reply:
I think this should be fixed in 21426412.